include/hip/hcc_detail/program_state.hpp

```diff
@@ -558,92 +561,166 @@ const std::unordered_map<
+void create_string_from_string_node_and_destroy(
+    amd_comgr_metadata_node_t printNode,
+    std::string& value) {
+    amd_comgr_status_t status;
+    size_t printSize;
+    char *printString;
 
-        static constexpr size_t align_sz{6};
-        f = metadata.find("Align:", f) + align_sz;
+    status = amd_comgr_get_metadata_string(printNode, &printSize, NULL);
+    checkError(status, "amd_comgr_get_metadata_string");
+    printString = (char*)malloc(printSize);
 
-        char* l{};
-        auto align = std::strtoul(&metadata[f], &l, 10);
+    status = amd_comgr_get_metadata_string(printNode, &printSize, printString);
+    checkError(status, "amd_comgr_get_metadata_string");
 
-        f += (l - &metadata[f]) + 1;
+    value = printString;
+    free(printString);
 
-        size_align.emplace_back(size, align);
-    } while (true);
+    status = amd_comgr_destroy_metadata(printNode);
+    checkError(status, "amd_comgr_destroy_metadata");
 }
 
 inline
-void read_kernarg_metadata(
-    ELFIO::elfio& reader,
+void process_kernarg_metadata(
+    amd_comgr_metadata_node_t blob_meta,
     std::unordered_map<
         std::string,
         std::vector<std::pair<std::size_t, std::size_t>>>& kernargs) {
-    // TODO: this is inefficient.
-    auto it = find_section_if(reader, [](const ELFIO::section* x) {
-        return x->get_type() == SHT_NOTE;
-    });
-
-    if (!it) return;
-
-    const ELFIO::note_section_accessor acc{reader, it};
-    for (decltype(acc.get_notes_num()) i = 0; i != acc.get_notes_num(); ++i) {
-        ELFIO::Elf_Word type{};
-        std::string name{};
-        void* desc{};
-        ELFIO::Elf_Word desc_size{};
+    amd_comgr_status_t status;
+    amd_comgr_metadata_kind_t mkindLookup;
+    amd_comgr_metadata_node_t kernelList, kernelMap;
+    amd_comgr_metadata_node_t kernArgList, kernArgMap;
+    amd_comgr_metadata_node_t printNode;
+    std::string printString;
+    size_t num_kernels = 0;
+    size_t num_kern_args = 0;
+
+    // Kernels is a list of MAPS!!
+    status = amd_comgr_metadata_lookup(blob_meta, "Kernels", &kernelList);
+    if(status != AMD_COMGR_STATUS_SUCCESS)
+        return;
+
+    status = amd_comgr_get_metadata_kind(kernelList, &mkindLookup);
+    if (mkindLookup != AMD_COMGR_METADATA_KIND_LIST) {
+        hip_throw(std::runtime_error{"Lookup of Kernels didn't return a list\n"});
+    }
 
-        acc.get_note(i, type, name, desc, desc_size);
+    status = amd_comgr_get_metadata_list_size(kernelList, &num_kernels);
+    checkError(status, "amd_comgr_get_metadata_list_size");
+    for (int i = 0; i < num_kernels; i++) {
+        status = amd_comgr_index_list_metadata(kernelList, i, &kernelMap);
+        checkError(status, "amd_comgr_index_list_metadata");
+
+        std::string kernel_name;
+        status = amd_comgr_metadata_lookup(kernelMap, "Name", &printNode);
+        checkError(status, "amd_comgr_metadata_lookup");
+        create_string_from_string_node_and_destroy(printNode, kernel_name);
```

Comment:
I don't find hiding the extraction of the string into this function improves readability.  Actually, the fact that it would destroy the string node makes the code even more obscure.

Reply:
Do you suggest to put the contents of `create_string_from_string_node_and_destroy` into each instance that I grab a string?

Reply:
or move the construction of the `std::string` and the metadata lookup into that this helper function (and rename it)?